lib/jeff.ex

```diff
@@ -174,6 +176,79 @@ defmodule Jeff do
     ACU.send_command(acu, address, MFG, params) |> handle_reply()
   end
 
+  @doc """
+  Send file data to a PD
+  """
+  @spec file_transfer(acu(), osdp_address(), binary()) ::
+          Reply.FileTransferStatus.t() | cmd_err()
+  def file_transfer(acu, address, data) when is_binary(data) do
+    file_transfer(acu, address, data, nil)
+  end
+
+  @spec file_transfer(acu(), osdp_address(), binary(), function() | nil) ::
+          Reply.FileTransferStatus.t() | cmd_err()
+  def file_transfer(acu, address, data, progress_callback) when is_binary(data) do
+    total_packets = div(byte_size(data) + @osdp_max_packet_size - 1, @osdp_max_packet_size)
+
+    FileTransfer.command_set(data, @osdp_max_packet_size)
+    |> run_file_transfer(acu, address, progress_callback, 0, total_packets)
+  end
+
+  defp run_file_transfer([cmd | rem], acu, address, progress_callback, packet_num, total_packets) do
+    case ACU.send_command(acu, address, FILETRANSFER, Map.to_list(cmd)) do
+      {:ok, reply} ->
+        handle_file_transfer_reply(
+          reply,
+          rem,
+          acu,
+          address,
+          progress_callback,
+          packet_num,
+          total_packets
+        )
+
+      error ->
+        error
+    end
+  end
+
+  defp run_file_transfer([], _acu, _address, progress_callback, _packet_num, total_packets) do
+    # Handle empty list case
+    if progress_callback do
+      progress_callback.(total_packets, total_packets, 100)
+    end
+
+    %Reply.FileTransferStatus{status: :ok}
+  end
+
+  defp handle_file_transfer_reply(
+         reply,
+         rem,
+         acu,
+         address,
+         progress_callback,
+         packet_num,
+         total_packets
+       ) do
+    case FileTransfer.adjust_from_reply(reply, rem) do
+      {:cont, next, delay} ->
+        call_progress_callback(progress_callback, packet_num + 1, total_packets)
+        :timer.sleep(delay)
+        run_file_transfer(next, acu, address, progress_callback, packet_num + 1, total_packets)
+
+      {:halt, data} ->
+        call_progress_callback(progress_callback, total_packets, total_packets)
+        data
+    end
+  end
+
+  defp call_progress_callback(nil, _packet_num, _total_packets), do: :ok
+
+  defp call_progress_callback(progress_callback, packet_num, total_packets) do
+    progress_percentage = round(packet_num * 100 / total_packets)
+    progress_callback.(packet_num, total_packets, progress_percentage)
+  end
```

Comment:
```suggestion
  defp call_progress_callback(progress_callback, packet_num, total_packets) when is_function(progress_callback, 3) do
    progress_percentage = round(packet_num * 100 / total_packets)
    progress_callback.(packet_num, total_packets, progress_percentage)
  end
  
  defp call_progress_callback(_cb, _packet_num, _total_packets), do: :ok
```

One way to better protect would be to guard on the callback actually being a 3 arity function and ignore anything else. I am indifferent on needing to catch anything the call might produce...maybe leave for now since if you provide a callback for progress, you probably don't want it silently passing?